Feature add repair requests

peterhsieh opened the pull request:
# In this PR

**Repair Requests**
- Created function for repair request calls

**Repair Requests Form**
- Created form for submitting repair requests

**API**
- Created Repair Requests API (addRepairRequest)

---
# Issues

The following issues involve further design of the application, hence I didn't go forward in making the changes.
1. Date object from DatePicker is not received by database correctly
   Currently the database 'date' for repair requests is set to TIMESTAMP. Discrepancy between frontend date payload -> actual date value in database.
2. Error occurs (Warning: flattenChildren(...): Encountered two children with the same key) when two repair requests have the same date. It appears the date is acting as a 'key' and causes only the oldest repair request to be visible (overriding the rest).

**Possible Solution**: Would it be more appropriate to assume that you cannot change the date of an initial submission of the repair request? Utilise the timestamp as the 'key'?
1. Repair Request Content
   What should actually be reported in a repair request? Date of submission (by default via database Timestamp - no payload), Type (e.g. dropdown of areas of repair - kitchen, bathroom, garage), description, date of incident?, photos, status?

peterhsieh committed:
(RepairRequest) added form, database link set, a few bugs to fix

peterhsieh committed:
(Repair Requests) separated form from repair request, working

nirrek commented:
To solve the key thing, you need to change this line:
https://github.com/nirrek/wallfly/blob/feature-add-repair-requests/scripts/components/RepairRequest.jsx#L43

The key (just like in a database) will be used to identify the entity. So this will need to be something unique, if we are assuming making multiple repair requests on a given day (which is probably a safe assumption to make). So just synthesize some key that will be unique.

Again, the place with all the info is the React documentation:
https://facebook.github.io/react/docs/multiple-components.html#dynamic-children

Regarding the database timestamp thing, I think we should just drop the time selection on the client side. You shouldn't get to decide when you lodge the request, as you could do stupid stuff like backdate it 2 weeks, and then be like `screw you agent you haven't got back to me in 2 weeks`. Anyways, this also fixes your problem as you can have the DB generate the timestamp based upon insertion time into the table, this is actually the default way that you use timestamps with DBs.

As to what should be in one. Lets just keep it bare minimum for the moment and we can add more as necessary. I think minimum required stuff is `timestamp`, `description`, `photo`, `status`. There will want to be like a comment thread on the request and other stuff, but lets not concern ourselves with that stuff yet.

peterhsieh committed:
Merge branch 'master' into feature-add-repair-requests

peterhsieh commented:
Alright! Yep I agree! I'll work on those changes.

peterhsieh committed:
(Repair Requests) changed structure of repair request

- Removed all instances of a Date for the submission of a repair
requests (date is generated by the database via TIMESTAMP)
- Removed subject from repair requests

nirrek committed:
(RepairRequestForm) make input a textarea, clearnup styles

nirrek committed:
(RepairRequest) Hide form by default. Handle no image in result.

nirrek commented:
Seems good. Shall merge.
